### PR TITLE
React api

### PR DESCRIPTION
**Pull request by makhitr**

### CODE REVIEW

**Comment by ValtsS** [comment]
Missing PR description

**Comment by makhitr** [comment]
1. Task: https://github.com/rolling-scopes-school/tasks/tree/master/react/modules/module04
2. Screenshot: 
изображение
3. Deploy https://makhitr-react-api.netlify.app/
4. Done 10/04/2023 / Deadline 10/04/2023
5. Score

File: src/components/searchBar/SearchBar.module.css
Comment (ValtsS) [inline src/components/searchBar/SearchBar.module.css:1]: Unexpected shorthand "margin" after "margin-bottom"

File: src/components/fullCard/FullCard.tsx [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  const [cardData, setCardData] = React.useState<ICard | null>(null);
+
+  React.useEffect(() => {
+    fetch(`https://rickandmortyapi.com/api/character/${id}`)
+      .then((response) => response.json())
+      .then((data) => setCardData(data));
+  }, [id]);
+  return (
+    <>
+      {cardData && (
```
Comment (ValtsS) [inline src/components/fullCard/FullCard.tsx:15]: Why do you have duplicate code? 23 lines starting from here and in Card.tsx

File: src/components/card/Card.module.css [hunk 2]
```diff
@@ -1,8 +1,18 @@
...
+  margin-bottom: 1.5rem;
+}
+.cardMain {
+  display: flex;
+  gap: 2rem;
+}
 .cardWrapper img {
-  width: 200px;
+  max-width: 50%;
+  border: 2px solid rgb(255,255,0);;
```
Comment (ValtsS) [inline src/components/card/Card.module.css]: double semicolon

File: src/components/cardsList/CardsList.tsx [hunk 3]
```diff
@@ -1,15 +1,42 @@
...
+            {selectedCardId && <FullCard id={selectedCardId} />}
+          </div>
+        </Modal>
+      )}
+      <div className={styles.cardSection} data-testid="cards-list">
+        {cards.map(
+          (card, id) =>
+            'location' in card && (
+              <CardPreview
+                key={`card.name-${id}`}
```
Comment (ValtsS) [inline src/components/cardsList/CardsList.tsx]: Do not use Array index in keys

File: src/components/fullCard/FullCard.module.css [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+.cardHeader {
+  margin-bottom: 1.5rem;
+}
+.cardMain {
+  display: flex;
+  gap: 2rem;
+}
+.cardWrapper img {
+  max-width: 50%;
+  border: 2px solid rgb(255,255,0);;
```
Comment (ValtsS) [inline src/components/fullCard/FullCard.module.css]: extra semilcolon, duplicate code

File: src/components/fullCard/FullCard.test.tsx [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+      });
+    },
+  });
+}) as Mock;
+describe('one Card test', () => {
+  it('render card', async () => {
+    await act(async () => {
+      render(<FullCard id={1} />);
+    });
+    // render(<FullCard id={1} />);
```
Comment (ValtsS) [inline src/components/fullCard/FullCard.test.tsx:40]: Remove this commented out code

File: src/pages/MainPage/MainPage.tsx [hunk 6]
```diff
@@ -1,22 +1,70 @@
...
-      {listItems.length && (
-        <div className={styles.cardSection} data-testid="cards-list">
-          {listItems}
-        </div>
+      <SearchBar search={handleSearch} />
+      {error ? (
+        <p>No information is available for a page</p>
+      ) : !isLoaded ? (
+        <Spinner />
+      ) : (
```
Comment (ValtsS) [inline src/pages/MainPage/MainPage.tsx:56]: extract this logic out of render

File: src/components/cardsList/CardsList.test.tsx [hunk 7]
```diff
@@ -1,54 +1,60 @@
...
+        created: '2017-11-04T18:50:21.651Z',
       },
     ];
     render(<CardsList cards={cards} />);
     const cardsList = screen.getByTestId('cards-list');
-    const allCards = screen.getAllByTestId('card');
+    const allCards = screen.getAllByTestId('card-preview');
     expect(cardsList).toBeInTheDocument();
     expect(allCards.length).toBe(2);
   });
```
Comment (ValtsS) [inline src/components/cardsList/CardsList.test.tsx:60]: Lacking tests for functionality besides default render

File: src/components/modal/Modal.tsx [hunk 8]
```diff
@@ -0,0 +1,41 @@
+import { useEffect, useRef, ReactNode } from 'react';
+import { createPortal } from 'react-dom';
+import styles from '../CardsList/CardsList.module.css';
+
+const modalRoot = document.createElement('div');
+modalRoot.setAttribute('id', 'modal-root');
+document.body.appendChild(modalRoot);
```
Comment (ValtsS) [inline src/components/modal/Modal.tsx]: This -15 points.

File: src/pages/MainPage/MainPage.tsx [hunk 9]
```diff
@@ -1,22 +1,70 @@
...
+  const [isLoaded, setIsLoaded] = React.useState(false);
+  const [cardData, setCardData] = React.useState([]);
+
+  React.useEffect(() => {
+    const value = localStorage.getItem('searchValue');
+    const url = value
+      ? `https://rickandmortyapi.com/api/character/?name=${value}`
+      : 'https://rickandmortyapi.com/api/character';
+
+    const fetchData = async () => {
```
Comment (ValtsS) [inline src/pages/MainPage/MainPage.tsx]: single responsibility principle broken. Seems also like a duplicate code

File: src/pages/MainPage/MainPage.tsx [hunk 10]
```diff
@@ -1,22 +1,70 @@
...
+        });
+    };
+
+    fetchData();
+  }, []);
+
+  const handleSearch = (value: string) => {
+    setIsLoaded(false);
+    setError(null);
+    fetch(`https://rickandmortyapi.com/api/character/?name=${value}`)
```
Comment (ValtsS) [inline src/pages/MainPage/MainPage.tsx]: and again.

File: src/components/modal/Modal.tsx (hunk 8 above)
Comment (makhitr) [reply src/components/modal/Modal.tsx]: fixed with document.body this manipulation is allowed
image

File: src/pages/MainPage/MainPage.tsx (hunk 9 above)
Comment (makhitr) [reply src/pages/MainPage/MainPage.tsx]: +

**Pull request merged**